### ISSUE TITLE
Update Mio

### ISSUE THREAD

**Issue by Thomasdezeeuw** [alacritty/alacritty#6486]
It seems alacritty_terminal is still using Mio v0.6, which isn't supported any more. Any thing I can do to help move it to Mio v0.8?

**Comment by kchibisov** [comment]
I think there were issues wrt signal hook and level triggering. If you're experienced with mio you can give it a try at porting.

**Comment by Thomasdezeeuw** [comment]
> I think there were issues wrt signal hook and level triggering.

Changing from level triggers to edge triggers (what Mio v0.7 and Mio v0.8 use) is usually done by putting a loop around it. The only difference is that you have to [drain readiness](https://docs.rs/mio/latest/mio/struct.Poll.html#draining-readiness), which is done by performing I/O until it returns a [`WouldBlock`](https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.WouldBlock) "error", others the OS returns no more events for the I/O source.

Signals are a bit trickier, I made [mio-signals](https://crates.io/crates/mio-signals) to support it, though it has a minimal style API. However it does have some caveats, mainly no Windows support (but contributions are welcome) and it setup has to be done between starting new threads (otherwise they will receive signals as normal).

> If you're experienced with mio you can give it a try at porting.

I maintain Mio ;)

**Comment by chrisduerr** [comment]
There's just very little reason for Alacritty to change considering the newer APIs are just worse for us. If someone wanted to update it they'd be welcome to try of course, but I'm not planning to.

**Comment by Thomasdezeeuw** [comment]
> There's just very little reason for Alacritty to change considering the newer APIs are just worse for us.

Can you elaborate on how the APIs are worse?

**Comment by chrisduerr** [comment]
I mean you kinda just did. It requires additional work ("usually done by putting a loop around it") and there's no Windows support for the signals library you've mentioned (though I don't think we need signals on Windows).

What exists works without requiring any modification and there are no issues with it as far as I'm aware.

**Comment by Thomasdezeeuw** [comment]
> I mean you kinda just did. It requires additional work ("usually done by putting a loop around it") and

True, but I'm afraid the level and oneshot trigger didn't work properly cross platform, see https://github.com/tokio-rs/mio/issues/928 for more details, so it's not without reason. 

> there's no Windows support for the signals library you've mentioned (though I don't think we need signals on Windows).

To be fair Mio v0.6 never supported signals to begin with.

> What exists works without requiring any modification and there are no issues with it as far as I'm aware.

I'm not convinced that is true, the level triggers could be broken, see https://github.com/tokio-rs/mio/issues/928.

But if you're not willing to update that's always an option. I just wanted to make sure that the maintainers no that we're no longer supporting Mio v0.6.

**Comment by kchibisov** [comment]
Just by looking at it we use `level` only for signal handling, and for the rest we use `Edge | Oneshot`. From what I can see moving to just `Edge` is trivial, given that `Oneshot` means that you need to manually rearm.

As for `level` for signals I'm not quite sure, but @Thomasdezeeuw you might have a suggestion how to handle that?

https://github.com/alacritty/alacritty/blob/778558060243505d07b0850a115e3944862db5cd/alacritty_terminal/src/event_loop.rs#L321

The relevant code starting point.

We also need anon pipes for windows.

**Comment by Thomasdezeeuw** [comment]
> Just by looking at it we use `level` only for signal handling, and for the rest we use `Edge | Oneshot`. From what I can see moving to just `Edge` is trivial, given that `Oneshot` means that you need to manually rearm, so I'd assume the way you do so is by removing and reinserting?

If you want to emulate the oneshot behaviour you'll indeed need to deregister and reregister. But may I ask is the use case for oneshot? With edge triggers you'll only be informed once of relevant I/O changes (e.g. new data to read) and not again until you've [drained the readiness](https://docs.rs/mio/latest/mio/struct.Poll.html#draining-readiness) (e.g. read all bytes currently available for reading).

> As for `level` for signals I'm not quite sure, but @Thomasdezeeuw you might have a suggestion how to handle that?
> 
> https://github.com/alacritty/alacritty/blob/778558060243505d07b0850a115e3944862db5cd/alacritty_terminal/src/event_loop.rs#L321
> 
> The relevant code starting point.

I looks like this a wrapper around `mio_extras`'s channel. This is trickier to replace directly. I would either use `std::sync::mpsc` or crossbeam's channel (which just got merged into `std::sync::mpsc`) for the channel implementation. Then to wake up another thread you'll need to use `mio::Waker` to create Mio events. I've implemented this before here: https://github.com/Thomasdezeeuw/heph/blob/34a8adc74c32572c8b56d6fef3053161f32953c3/rt/src/channel.rs, feel free the copy that implementation.

> We also need anon pipes for windows.

I don't believe we can support Windows's anonymous pipes, though I'm a little fuzzy on the details, see https://github.com/tokio-rs/mio/issues/321 and https://github.com/tokio-rs/mio/issues/320. We do support named pipes: https://docs.rs/mio/latest/x86_64-pc-windows-msvc/mio/windows/struct.NamedPipe.html.

**Comment by Thomasdezeeuw** [comment]
> That's because we change `interest` between polling. So it's R and sometimes RW. So that's why we have oneshot, since that's what we want.

Have you tried registering with RW and edge triggers (default in Mio v0.7+) and then simply attempting the I/O operation when required, without re/de-registering in between?

> Though I think you can alter interest on the fly without oneshot?

Yes, using `reregister`.

> Yet not sure how to handle signals, I'm not that familiar with egde triggering though, since the default on linux is level...

The difference between edge and level triggers is that level triggers always return a poll event when I/O is ready, e.g. when data is ready to be read. Edge triggers on the other hand will tell you once that the operation is ready, then you have to process (e.g. read) it all before it returns another poll event. When edge triggers are used properly it's actually more efficient.

> Disregard anon pipes, we use miow for them, which is not related to mio, so it's irrelevant.

👍

**Comment by BearOve** [comment]
@chrisduerr you may want to revisit this as it has been "time bombed" by rust. When compiling on windows you now get `warning: the following packages contain code that will be rejected by a future version of Rust: winapi v0.2.8` . This version of winapi is what is used by mio 0.6 and since that is unsupported I guess it won't update winapi.

**Comment by kchibisov** [comment]
We need to reimplement lots of code on windows to make stuff work, and given that we have no contributors on windows it could take time. It's not like we can't fork old mio/miow/anon-pipes and migrate them to newer winapi/windows-sys. It's not the end of the world.

**Comment by kchibisov** [comment]
We decided to remove `mio` dependency in favor of `polling`, given that `polling` will be used as a part of `winit` as well for all the platforms(less dependencies in general).

See https://github.com/alacritty/alacritty/pull/6846
